packages-next/fields-document/ARCHITECTURE.md

```diff
@@ -0,0 +1,107 @@
+
+- Renderers for elements/leaves should never use `useSlate` or `useToolbarState` because those hooks will cause a re-render on every editor change. If they need to get the editor to do something in response to a user action, they should use `useStaticEditor` which will not cause re-renders.
+- All state derived from the document/selection that is shown in the toolbar should come from `useToolbarState` so that it is only computed once
+- Components that use `useToolbarState` and are shown when the user is typing (i.e. things that aren't in dialogs) should memoize the React elements they return with `useMemo` to avoid re-rendering when the state they consume from `useToolbarState` hasn't changed
+- Calls to `useToolbarState` should be as deep as possible to re-render the smallest number of components change when the relevant part of the toolbar state changes (e.g. a tooltip should be above the component that renders the tooltip so that it doesn't need to re-render even if the button inside it needs to re-render)
+
+## Normalization
+
+Normalization is used in the editor to enforce a specific structure.
+Slate runs normalization by going through all the changed or "dirty" nodes when changes happen on client and calling `editor.normalizeNode` from the deepest node that changed up to the `Editor` until the normalization doesn't make any more changes.
+The entire document is also normalized in `resolveInput`.
+
+For example, if you have two lists next to each other, you want to merge them so that they're one list.
+You don't want to do this in response to a specific user action because any number of things could result in two lists being next to each other so this is where you would use normalization.
+
+Besides bespoke normalization for the different types of nodes, there is a set of generic normalization which enforces which block types can have which children. Search for `editorSchema` in the to see how this works.
+
+When a node is moved, normalization will not happen on that node because it did not actually change.
+Normalization will happen on the old and the new parent nodes of the node that moved because both of the parent nodes children arrays changed.
+If you want to enforce something about the relationship between a parent and child node, you must do this by doing normalization when normalizeNode is passed the parent node since the child node won't be passed to normalizeNode when it is moved.
+
+## User Input
+
+Some behaviour should happen in response to user input, for example markdown shortcuts.
+These shouldn't happen in normalization because you still want people to be able to write `> ` for example without it becoming a blockquote by undoing it so for these kind of things, they're handled by overriding `editor.insertText`, `editor.insertBreak`, `editor.insertData` or `onKeyDown` (which is a prop to the `Editable` component rather than on the editor instance).
```

Comment:
This sentence is bit long and confusing, I'm not quite sure what it's trying to say. Could you rephrase it?